File: ts-packages/qgis-node/bin/qgis-plugin.js
#!/usr/bin/env node
/**
 * qgis-plugin — Node.js wrapper for Rust-native qgis-plugin binary
 */

const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

function findBinary(name) {
  const possiblePaths = [
    path.join(__dirname, '..', 'target', 'release', name),
    path.join(__dirname, '..', '..', '..', 'target', 'release', name),
    path.join(__dirname, name),
  ];
  const cargoHome = process.env.CARGO_HOME || path.join(require('os').homedir(), '.cargo', 'bin');
  possiblePaths.push(path.join(cargoHome, name));

  try {
    const { execSync } = require('child_process');
    const which = process.platform === 'win32' ? 'where' : 'which';
    const result = execSync(`${which} ${name}`, { encoding: 'utf-8', stdio: ['ignore', 'pipe', 'ignore'] });
    const found = result.trim().split('\n')[0];
    if (found) possiblePaths.unshift(found);
  } catch (_) {}

  for (const p of possiblePaths) {
    if (fs.existsSync(p)) return p;
    if (process.platform === 'win32' && fs.existsSync(p + '.exe')) return p + '.exe';
  }
  return null;
}

function main() {
  const args = process.argv.slice(2);

  if (args.length === 0 || args.includes('--help') || args.includes('-h')) {
    console.log('qgis-plugin — QGIS plugin SDK (Rust-native, from qgis-rs npm)');
    console.log('');
    console.log('Usage: qgis-plugin <command> [options]');
    console.log('');
    console.log('Commands:');
    console.log('  new       Scaffold a new plugin');
    console.log('  build     Build plugin');
    console.log('  test      Run tests');
    console.log('  install   Install into QGIS');
    console.log('  dev       Watch mode');
    console.log('  package   Create .zip');
    console.log('  validate  Check structure');
    console.log('  info      Show plugin info');
    console.log('  version   Print version');
    console.log('');
    return 0;
  }

  const binary = findBinary('qgis-plugin');
  if (binary) {
    const child = spawn(binary, args, { stdio: 'inherit' });
    child.on('exit', (code) => process.exit(code || 0));
    child.on('error', () => {
      console.error('qgis-plugin: binary failed, falling back to Python CLI if available');
      process.exit(1);
    });
  } else {
    console.log(`qgis-plugin: Rust binary not found, trying Python fallback...`);
    // Try Python CLI
    const { spawn: pySpawn } = require('child_process');
    const pyChild = pySpawn('python', ['-m', 'qgis_sdk.cli', ...args], { stdio: 'inherit' });
    pyChild.on('exit', (code) => process.exit(code || 0));
    pyChild.on('error', () => {
      console.error('qgis-plugin: no binary found. Install via: npm install qgis-rs or pip install qgis-sdk');
      process.exit(1);
    });
  }
}

main();

File: packages/qgis-node/src/bin/qgis-plugin.rs
//! qgis-plugin binary for npm package

use std::process::ExitCode;
use clap::Parser;

#[derive(Debug, Parser)]
#[command(name = "qgis-plugin", version, about = "QGIS plugin SDK — native speed")]
struct Cli {
    #[arg(long)]
    version: bool,
    #[arg(trailing_var_arg = true, allow_hyphen_values = true)]
    args: Vec<String>,
}

fn main() -> ExitCode {
    // For npm package, we delegate to qgis-sdk crate's CLI if available
    // Simplified: just print version and help, actual logic in qgis-sdk crate
    let args: Vec<String> = std::env::args().collect();
    if args.len() > 1 && (args[1] == "--version" || args[1] == "-V" || args[1] == "version") {
        println!("qgis-plugin {} (from qgis-rs npm, Rust-native)", env!("CARGO_PKG_VERSION"));
        return ExitCode::SUCCESS;
    }
    
    // Try to find qgis-sdk binary logic — for now, just show help
    if args.len() == 1 || args.iter().any(|a| a == "--help" || a == "-h") {
        println!("qgis-plugin {} — QGIS plugin SDK (Rust-native)", env!("CARGO_PKG_VERSION"));
        println!();
        println!("Usage: qgis-plugin <command> [options]");
        println!();
        println!("Commands:");
        println!("  new       Scaffold a new plugin");
        println!("  build     Build plugin");
        println!("  test      Run tests");
        println!("  install   Install into QGIS");
        println!("  dev       Watch mode");
        println!("  package   Create .zip");
        println!("  validate  Check structure");
        println!("  info      Show plugin info");
        println!("  version   Print version");
        println!();
        println!("For full CLI, install qgis-sdk Python package or use npx qgis-rs");
        return ExitCode::SUCCESS;
    }
    
    eprintln!("qgis-plugin: full functionality requires qgis-sdk Rust crate");
    eprintln!("qgis-plugin: install via: npm install qgis-rs && npx qgis-plugin --help");
    ExitCode::FAILURE
}
